core/src/main/java/io/methvin/watchservice/MacOSXListeningWatchService.java

```diff
@@ -254,11 +286,21 @@ public void invoke(
           }
         }
 
-        for (Path file : findDeletedFiles(fileName, filesOnDisk)) {
+        List<Path>  deletedPaths = findDeletedFiles(fileName, filesOnDisk);
+
+        if (hashCodeMap.isEmpty()) {
+          // all underlying paths are gone, so stop this service and cancel the key
+          watchKey.watchService().close(runLoopThread, this, Paths.get(absPath));
```

Comment:
From JDK, note item two:
A watch key is created when a watchable object is registered with a watch service. The key remains valid until:

-It is cancelled, explicitly, by invoking its cancel method, or
-Cancelled implicitly, because the object is no longer accessible, or
-By closing the watch service.

Reply:
You can look at the polling JDK impl
https://github.com/frohoff/jdk8u-jdk/blob/master/src/share/classes/sun/nio/fs/PollingWatchService.java#L326
'''            DirectoryStream<Path> stream = null;
            try {
                stream = Files.newDirectoryStream(watchable());
            } catch (IOException x) {
                // directory is no longer accessible so cancel key
                cancel();
                signal();
                return;
            }'''